src/ert/com_protocol/_state_machine.py

```diff
@@ -0,0 +1,233 @@
+        ensemble: pb2.Ensemble = self.get_ensemble(_id.ensemble)
+        if _id.realization not in ensemble.realizations:
+            ensemble.realizations[_id.realization].CopyFrom(pb2.Realization(id=_id))
+        return ensemble.realizations[_id.realization]
+
+    def get_step(self, _id: pb2.StepId) -> pb2.Step:
+        logger.debug(f"step id: {_id}")
+        real: pb2.Realization = self.get_realization(_id.realization)
+        if _id.step not in real.steps:
+            real.steps[_id.step].CopyFrom(pb2.Step(id=_id))
+        return real.steps[_id.step]
+
+    def get_job(self, _id: pb2.JobId) -> pb2.Job:
+        logger.debug(f"job id: {_id}")
+        step: pb2.Step = self.get_step(_id.step)
+        if _id.index not in step.jobs:
+            step.jobs[_id.index].CopyFrom(pb2.Job(id=_id))
+        return step.jobs[_id.index]
+
+
+class ExperimentStateMachine:
+    """The :class:`ExperimentStateMachine` implements a state machine for the
+    entire experiment. It allows an experiment to track its own state and
+    communicate it to others."""
+
+    def __init__(self) -> None:
+        self._state_handle: _StateHandle = _StateHandle()
+
+    def successful_realizations(self, ensemble_id: str) -> int:
+        """Return an integer indicating the number of successful realizations
+        in the experiment given its ensemble ``id``.
+        return 0 if the ensemble has no successful realizations or ensemble has
+        no realizations registered."""
+
+        if ensemble_id not in self.state.ensembles:
+            return 0
+        return sum(
+            self.state.ensembles[ensemble_id].realizations[id_real].status
+            == pb2.STEP_SUCCESS
+            for id_real in self.state.ensembles[ensemble_id].realizations
+        )
+
+    class IllegalStateUpdate(Exception):
+        def __init__(self, reason: str):
+            super().__init__(reason)
+            self.reason: str = reason
+
+    @singledispatchmethod
+    def _update(
+        self,
+        _: Union[pb2.Job, pb2.Step, pb2.Realization, pb2.Ensemble, pb2.Experiment],
+    ) -> None:
+        pass
+
+    @_update.register
+    def _(self, job: pb2.Job) -> None:
+        logger.debug(
+            "Updating job id %d status %s from pbuf!",
+            job.id.index,
+            pb2.JobStatus.Name(job.status),
+        )
+        old_job = self._state_handle.get_job(job.id)
+        old_job.MergeFrom(job)
+
+    @_update.register
+    def _(self, real: pb2.Realization) -> None:
+        logger.debug(f"Updating realization id {real.id.realization} from pbuf!")
+        old_real = self._state_handle.get_realization(real.id)
+        old_real.MergeFrom(real)
+
+    @_update.register
+    def _(self, step: pb2.Step) -> None:
+        logger.debug(
+            f"Updating step id {step.id.step} status "
+            f"{pb2.StepStatus.Name(step.status)} from pbuf!",
+        )
+        old_real = self._state_handle.get_step(step.id)
```

Comment:
`old_real`? (wrong reusage of variable name?)

Reply:
Good catch!